src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -1279,177 +1105,146 @@ protected ObjectMetadata putObject(
-        }
         if (objSize != -1) {
             FileHashStoreUtility.checkNotNegativeOrZero(objSize, "putObject");
         }
 
-        // If validation is desired, checksumAlgorithm and checksum must both be present
-        boolean requestValidation = verifyChecksumParameters(checksum, checksumAlgorithm);
-
         // Generate tmp file and write to it
-        logFileHashStore.debug("FileHashStore.putObject - Generating tmpFile");
         File tmpFile = FileHashStoreUtility.generateTmpFile("tmp", OBJECT_TMP_FILE_DIRECTORY);
-        Path tmpFilePath = tmpFile.toPath();
         Map<String, String> hexDigests;
         try {
             hexDigests = writeToTmpFileAndGenerateChecksums(
                 tmpFile, object, additionalAlgorithm, checksumAlgorithm
             );
         } catch (Exception ge) {
             // If the process to write to the tmpFile is interrupted for any reason,
-            // we will delete the tmpFile. 
-            boolean deleteStatus = tmpFile.delete();
+            // we will delete the tmpFile.
+            Files.delete(tmpFile.toPath());
             String errMsg =
-                "FileHashStore.putObject - Unexpected Exception while storing object for: " + pid;
-            if (deleteStatus) {
-                errMsg = errMsg + ". Deleting temp file: " + tmpFile + ". Aborting request.";
-            } else {
-                errMsg = errMsg + ". Failed to delete temp file: " + tmpFile
-                    + ". Aborting request.";
-            }
+                "Unexpected Exception while storing object for pid: " + pid + ". " + ge.getMessage();
             logFileHashStore.error(errMsg);
             throw new IOException(errMsg);
         }
-        long storedObjFileSize = Files.size(Paths.get(tmpFile.toString()));
 
         // Validate object if checksum and checksum algorithm is passed
         validateTmpObject(
-            requestValidation, checksum, checksumAlgorithm, tmpFilePath, hexDigests, objSize,
-            storedObjFileSize
-        );
+            compareChecksum, checksum, checksumAlgorithm, tmpFile, hexDigests, objSize);
 
         // Gather the elements to form the permanent address
         String objectCid = hexDigests.get(OBJECT_STORE_ALGORITHM);
-        String objShardString = FileHashStoreUtility.getHierarchicalPathString(
+        String objRelativePath = FileHashStoreUtility.getHierarchicalPathString(
             DIRECTORY_DEPTH, DIRECTORY_WIDTH, objectCid
         );
-        Path objRealPath = OBJECT_STORE_DIRECTORY.resolve(objShardString);
+        Path objRealPath = OBJECT_STORE_DIRECTORY.resolve(objRelativePath);
 
         // Confirm that the object does not yet exist, delete tmpFile if so
         if (Files.exists(objRealPath)) {
-            String errMsg = "FileHashStore.putObject - File already exists for pid: " + pid
-                + ". Object address: " + objRealPath + ". Deleting temporary file.";
+            Files.delete(tmpFile.toPath());
```

Comment:
When you check if the file exists on line 1151, do you need to lock the cid? If two threads store the same object with different pids, they could create a race condition and one thread may fail.

Reply:
Thank you for bringing this up - I believe in the process of adding the synchronization I discovered the true issue of my Runnable Test bug! I was able to produce that bug as I was writing the additional code - which now seems to be resolved.